### ISSUE TITLE
[iOS] RCT_EXTERN_MODULE Swift modules broken in Xcode 10.2

### ISSUE THREAD

**Issue by kdawgwilk** [facebook/react-native#24139]
## 🐛 Bug Report

It appears Obj-C runtime got more strict with Swift ABI stability in Xcode 10.2 and now causes a crash at runtime when trying to load modules that have Swift implementations.  This is the crash error:

```
Swift class extensions and categories on Swift classes are not allowed to have +load methods
```

## To Reproduce

Create a RN NativeModule e.g.

`BadgeHandler.m`
```objc
#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE(BadgeHandler, NSObject)
    RCT_EXTERN_METHOD(setBadgeNumber:(int *)badgeNumber)
@end
```

`BadgeHandler.swift`
```swift
import UIKit

@objc(BadgeHandler)
class BadgeHandler: NSObject, RCTBridgeModule {
    static func moduleName() -> String! {
        return "BadgeHandler";
    }

    static func requiresMainQueueSetup() -> Bool {
        return true
    }
    
    @objc func setBadgeNumber(_ badgeNumber: Int) {
        DispatchQueue.main.async {
            UIApplication.shared.applicationIconBadgeNumber = badgeNumber
        }
    }
}
```
`AppName-Bridging-Header.h`
```objc
#ifndef AppName_Bridging_Header_h
#define AppName_Bridging_Header_h

#import <React/RCTBridgeModule.h>

#endif /* AppName_Bridging_Header_h */
```

## Expected Behavior

The NativeModule would load fine at runtime like it did in previous versions of Xcode

## Code Example

See above ^

## Environment

```
  React Native Environment Info:
    System:
      OS: macOS 10.14.3
      CPU: (12) x64 Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
      Memory: 31.36 MB / 32.00 GB
      Shell: 5.3 - /bin/zsh
    Binaries:
      Node: 11.6.0 - /usr/local/bin/node
      Yarn: 1.13.0 - /usr/local/bin/yarn
      npm: 6.5.0 - /usr/local/bin/npm
      Watchman: 4.9.0 - /usr/local/bin/watchman
    SDKs:
      iOS SDK:
        Platforms: iOS 12.2, macOS 10.14, tvOS 12.2, watchOS 5.2
      Android SDK:
        API Levels: 23, 25, 26, 27, 28
        Build Tools: 27.0.3, 28.0.2, 28.0.3
        System Images: android-28 | Google APIs Intel x86 Atom
    IDEs:
      Xcode: 10.2/10E125 - /usr/bin/xcodebuild
    npmPackages:
      react: 16.8.3 => 16.8.3
      react-native: 0.59.1 => 0.59.1
```

## Other

This seems to point to this macro in `RCTBridgeModule.h`:

https://github.com/facebook/react-native/blob/d9eae2a8093156c8faae0c5ae14546bbc95e7a3d/React/Base/RCTBridgeModule.h#L71-L74

**Comment by callcter** [comment]
same issue

**Comment by djbird2046** [comment]
same issue, and i see the 0.59.2, + (void)load { RCTRegisterModule(self); }  , is not change

**Comment by JingGao1990** [comment]
Same issue here. Seeing this after xcode upgrade (10.2)

**Comment by alechangbiao** [comment]
Same here, upgrade to 10.2 and it appears

**Comment by juliandramirez** [comment]
same issue here

**Comment by mx-nagaokakoichi** [comment]
same issue

**Comment by farhad-ris** [comment]
same issue here

**Comment by tximpa91** [comment]
same issue here

**Comment by JackyLeeInChina** [comment]
same issue

**Comment by iqysf** [comment]
same issue

**Comment by quangbachit** [comment]
same issue

**Comment by brenwell** [comment]
Anyone got a temporary hack to get around this one?

**Comment by baveku** [comment]
same issue

**Comment by dcristolovean** [comment]
same here, this needs to be fixed asap, we're stuck because of this.

**Comment by solidfox** [comment]
The (semi) quick fix is to downgrade to Xcode 10.1.

**Comment by kelset** [comment]
Hey everyone - two things:

1) writing comments like "same issue" doesn't help at all. There are reactions for that. To prevent more spam, I'll lock the issue for now

2) if you want to help, ideal case would be to submit a PR to fix the issue. Feel free to @ me in the PR so that we can try to have it review'd and merged quickly.

3) Please make sure you are using RN 0.59.2 because that's the only version we are going to patch once the PR with the fix lands - at least for now. To upgrade please refer to the documentation -> https://facebook.github.io/react-native/docs/upgrading#react-native-projects (if you are upgrading from an earlier version than 0.59.0 refer to https://github.com/pvinis/rn-diff-purge)

We'll write more comments when we have more info on this.

**Comment by joostlubach** [comment]
I've implemented a (temporary) workaround, but I'm not sure how to fix it properly in React.

In your (Swift based) bridge modules, use the marco `RCT_EXPORT_PRE_REGISTERED_MODULE()` instead of `RCT_EXPORT_MODULE`. This prevents the `+ (void)load` method to be implemented on it.

Then, in your main.h, just register the Swift modules manually:

```
#import <UIKit/UIKit.h>
#import <React/RCTDefines.h>

#import "AppDelegate.h"
#import "<YourProjectName>-Swift.h"

RCT_EXTERN void RCTRegisterModule(Class);

int main(int argc, char * argv[]) {
  @autoreleasepool {
    RCTRegisterModule([Module1 self]);
    RCTRegisterModule([Module2 self]);

    return UIApplicationMain(argc, argv, nil, NSStringFromClass([AppDelegate class]));
  }
}
```

**Comment by arosendorff** [comment]
I found this while researching the issue, I'm not sure if it's relevant: https://github.com/pmusolino/Wormholy/pull/52

I tried to implement a similar solution, but my Objective-C is not very good, so I'm not sure how to call RCTRegisterModule(self) if we have to create a C/C++ function using __attribute__((constructor)). Perhaps someone smarter than I could work it out!

**Comment by zienag** [comment]
Hi! I'm not very familiar with react-native, but I found myself here because I'm blocked from upgrading to 10.2 by one of dependency that use react-native.

I have an idea for a fix though:
```
#define RCT_EXPORT_MODULE(js_name, objc_name) \
RCT_EXTERN void RCTRegisterModule(Class); \
+ (NSString *)moduleName { return @#js_name; } \
__attribute__((constructor)) \
static void RCT_CONCAT(initialize_, objc_name) { RCTRegisterModule([objc_name class]); }
```
The problem with this is that `RCT_EXPORT_MODULE` macro will require objc_name parameter now. As I understand, this is not very convenient, as I is used in many places by users. So another solution is to introduce another macro:
```
#define RCT_EXPORT_MODULE_NO_LOAD(js_name, objc_name) \
RCT_EXTERN void RCTRegisterModule(Class); \
+ (NSString *)moduleName { return @#js_name; } \
__attribute__((constructor)) \
static void RCT_CONCAT(initialize_, objc_name) { RCTRegisterModule([objc_name class]); }
```
And replace call in RCT_EXTERN_REMAP_MODULE to new macro:
```
#define RCT_EXTERN_REMAP_MODULE(js_name, objc_name, objc_supername) \
  objc_name : objc_supername \
  @end \
  @interface objc_name (RCTExternModule) <RCTBridgeModule> \
  @end \
  @implementation objc_name (RCTExternModule) \
  RCT_EXPORT_MODULE_NO_LOAD(js_name, objc_name)
```
I'm ok to make a pr, but it will take some time until I figure out how to setup everything.

**Comment by kelset** [comment]
@zienag a PR would be super appreciated :) I've tried to raise this with the FB team but they are all in US atm so it will take them a few more hours to reply to me.

**Comment by euroclydon37** [comment]
I'd like to clarify that I don't get this error on a physical device.

**Comment by euroclydon37** [comment]
Specifically, the error doesn't happen on devices/simulators prior to 12.2. So it doesn't seem xcode version related. It seems more tied to swift version.

**Comment by davethomas11** [comment]
If you can't update your react-native version to get this fix when it is merged because you are using the Expo fork of react-native. Then add this code ( based on above PR ) to a header file:

```
/**
 * Same as RCT_EXPORT_MODULE, but uses __attribute__((constructor)) for module
 * registration. Useful for registering swift classes that forbids use of load
 * Used in RCT_EXTERN_REMAP_MODULE
 */
#define RCT_EXPORT_MODULE_NO_LOAD(js_name, objc_name) \
RCT_EXTERN void RCTRegisterModule(Class); \
+ (NSString *)moduleName { return @#js_name; } \
__attribute__((constructor)) static void \
RCT_CONCAT(initialize_, objc_name)() { RCTRegisterModule([objc_name class]); }

#define RCT_EXTERN_MODULE_2(objc_name, objc_supername) \
RCT_EXTERN_REMAP_MODULE_2(, objc_name, objc_supername)

/**
 * Like RCT_EXTERN_MODULE, but allows setting a custom JavaScript name.
 */
#define RCT_EXTERN_REMAP_MODULE_2(js_name, objc_name, objc_supername) \
objc_name : objc_supername \
@end \
@interface objc_name (RCTExternModule) <RCTBridgeModule> \
@end \
@implementation objc_name (RCTExternModule) \
RCT_EXPORT_MODULE_NO_LOAD(js_name, objc_name)
```

Then in your module .m objective C file import that header file you just created and use `RCT_EXTERN_MODULE_2` instead of `RCT_EXTERN_MODULE`

Example:
```
#import "RctExternModule2HeaderFileThatYouCreatedReplaceFilenameHere.h"
#import "React/RCTViewManager.h"

@interface RCT_EXTERN_MODULE_2(SomeModuleNameOfYoursHere, RCTViewManager)
@end
```

**Comment by Tregorian** [comment]
My current workaround was to install iOS 12.1 simulator from xcode => preferences => components,
and after that just launched the app with older simulator.
It seems that the newer simulators use a lower-case second letter and the older ones capital.
for example Xs => XS

**Comment by euroclydon37** [comment]
Yeah, I had to rename my iPhone XR simulator. Apple had changed it to *iPhone Xʀ*. They used some special character that renders a smaller, yet still uppercase, letter. Fancy, Apple... fancy.

**Comment by omatrot** [comment]
Not only am I a module maker, but also a module user.
So, does this means that all swift modules out there will fail with XCode 10.2 ?
I can't see how patching RN will automatically fix third party modules. Am I wrong?

**Comment by arosendorff** [comment]
@omatrot zienag's proposed fix will automatically fix third part modules, because it modifies RCT_EXTERN_MODULE, which is how all Swift modules export themselves! If you simply include zienag's fix manually it will allow you to keep working using iOS 12.2.

Alternatively using a simulator for iOS 12.1 will cause this not to be an error as the issue is caused by the iOS version, not the XCode version 🙂

**Comment by omatrot** [comment]
I'm trying @zienag proposed fix, the crash is gone but there are a lot of warnings and the app does not work. For instance:

```
Class RNFetchBlob was not exported. Did you forget to use RCT_EXPORT_MODULE()?
````

Basically all the native modules are broken now.

**Comment by brenwell** [comment]
@omatrot I have applied the fix and am also using RNFetchBlob and everything is working exactly as expected, are you positive you have made the change correctly?

**Comment by omatrot** [comment]
@brenwell I've overwritten the whole node_modules/react-native/React/Base/RCTBridgeModule.h with one coming from the [pull request](https://github.com/facebook/react-native/pull/24155/commits/1eb1624cee5fea46eb11202297c13bc9a28d4131)...

**Comment by juliandramirez** [comment]
@omatrot zienag's fix worked for me as well. Maybe you are missing something else.
BTW I used https://github.com/ds300/patch-package to patch my react installation, I highly recommend it

**Comment by omatrot** [comment]
@brenwell  @juliandramirez  Are you using CocoaPods for your native dependencies?

**Comment by zienag** [comment]
> @omatrot zienag's proposed fix will automatically fix third part modules, because it modifies RCT_EXTERN_MODULE, which is how all Swift modules export themselves! If you simply include zienag's fix manually it will allow you to keep working using iOS 12.2.
> 
> Alternatively using a simulator for iOS 12.1 will cause this not to be an error as the issue is caused by the iOS version, not the XCode version 🙂

Actually this is not entirely true, as it doesn't change "binary code" in react-native itself, it patches code that generates from macros in users code. 
To fix be applied you need to recompile all the modules. That means, if you have dependencies that are distributing in binary form (static or dynamic lib) you need to recompile all dependencies with fix applied.
@omatrot do you have binary dependencies?

**Comment by omatrot** [comment]
@zienag I would say no, because my Podfile contains:
```
use_frameworks!
```

**Comment by zienag** [comment]
> @zienag I would say no, because my Podfile contains:
> 
> ```
> use_frameworks!
> ```

AFAIK, that is not related to this. Binary dependency means that it is prebuilt and there is no available code (aka "proprietary", for cocoapods it usually means that instead of code you have framework, For example, different ads/tracking libs are distributed like this, Fabric or Google Mobile Ads SDK). If you have prebuilt dependency that uses RCT_EXTERN_MODULE from react-native it won't work until recompiled with fix.

**Comment by omatrot** [comment]
@zienag Now I'm confused because it is not about RCT_EXTERN_MODULE but RCT_EXPORT_MODULE in the error messages. Are you sure the fix is not breaking this macro?

**Comment by zienag** [comment]
> @zienag Now I'm confused because it is not about RCT_EXTERN_MODULE but RCT_EXPORT_MODULE in the error messages. Are you sure the fix is not breaking this macro?

I see, I were wrong, sorry. Although issue that I described exists, that is not your case. It seems like `__attribute__((constructor))` is not called. Are you sure that you are not using `RCT_EXPORT_PRE_REGISTERED_MODULE` from previous workaround? If yes, can you run function with constructor attribute? Something like this, put it anywhere in your objc code, build and run, it should print log before entering `main`:
```
__attribute__((constructor)) static void justConstructor() {
  NSLog(@"hello from function with constructor attribute");
}
```

**Comment by omatrot** [comment]
@zienag I'm sure I'm not using the previous Workaround, because I wont patch third party modules. I'm using your modifications in RCTBridgeModule.h. I'm not experiencing crash at startup anymore but all these error messages about modules not exported.

**Comment by yassinecc** [comment]
> @brenwell I've overwritten the whole node_modules/react-native/React/Base/RCTBridgeModule.h with the one coming from the [pull request](https://github.com/facebook/react-native/pull/24155/commits/1eb1624cee5fea46eb11202297c13bc9a28d4131)...

@omatrot Which react-native version are you using? You might have issues if you use `RCTBridgeModule.h` from master against a previous version of RN

**Comment by omatrot** [comment]
I'm on RN 0.59.1.

**Comment by zienag** [comment]
Hmm, `RCTNativeAnimatedModule` doesn't even use `RCT_EXTERN_MODULE` macro. It uses `RCT_EXPORT_MODULE`, pr doesn't touch it at all.

**Comment by yassinecc** [comment]
@omatrot Can you try only adding @zienag 's changes to your `RCTBridgeModule.h`?

**Comment by omatrot** [comment]
@zienag I know this is weird.
@yassinecc This is what I've done. 

- Without the changes I'm experiencing crashes at startup with the same error highlighted in this issue.
- With the changes, no crash but none of the third party modules I'm using are exported anymore...

**Comment by omatrot** [comment]
I would say that we need to patch RCT_EXPORT_MODULE as well. What do you think?

**Comment by renzhezizhi** [comment]
I found a similar [issue](https://github.com/ionic-team/capacitor/issues/1262) in one of the iconic-team. They just deleted all "load" methods in their code to avoid this problem.
I searched inside the react-native folder and found THREE files with +(void)load method for ios. They are:
React/Base/RCTBridgeModule.h
Libraries/WebSocket/RCTReconnectingWebSocket.m
RNTester/Swizzle_RCTFabricSurface.m

Hope it helps.

**Comment by brenwell** [comment]
@omatrot yep using cocoapods on 0.59.2. I build for simulator, device and made a release. No problems

**Comment by kelset** [comment]
@renzhezizhi can you comment on this PR about your finding -> https://github.com/facebook/react-native/pull/24155

Because I think that PR is trying to fix the issue in a similar way 🤗

**Comment by zienag** [comment]
> I would say that we need to patch RCT_EXPORT_MODULE as well. What do you think?

I wrote small comment in pr about it. As addition, changing `RCT_EXPORT_MODULE` probably will not help you, unless we just discovered a serious bug in objc runtime. I think you can try two things:
1. Update react native to 59.2
2. If it will not help, try to run preprocessor in one of the files that are not registering. For example, given file:
```
MyModule.m:
@implementation MyModule
...
RCT_EXPORT_MODULE("MyModule")
...
@end
```
Running preprocessor gives something like this:
```
@implementation MyModule
...
extern __attribute__((visibility("default"))) void RCTRegisterModule(Class); 
+ (NSString *)moduleName { return @"MyModule"; } 
+ (void)load { 
  RCTRegisterModule(self); 
}
...
@end
```
If you compare code before and after fix – it should be different. If it is not, that is not related to fix from pr.

**Comment by omatrot** [comment]
@zienag I've done point 1 already. No change. I was thinking about running the preprocessor. I'll do that.

**Comment by Pranit-Harekar** [comment]
@davethomas11 solution helped me. Thanks!! 👍

**Comment by kelset** [comment]
(let's keep this open until the PR lands)

**Comment by charpeni** [comment]
Sorry about closing the issue here, looks like GitHub autoclose issues based on private references as well. 😖

**Comment by omatrot** [comment]
@zienag 

Here we go:

Message:
```
2019-03-27 19:34:11.944 [info][tid:main][RCTBridge.m:134] Class RNDeviceInfo was not exported. Did you forget to use RCT_EXPORT_MODULE()?
```
Preprocessor result on file :
```
extern __attribute__((visibility("default"))) void RCTRegisterModule(Class); + (NSString *)moduleName { return @"RNDeviceInfo"; } + (void)load { RCTRegisterModule(self); }
```
The files are compiled correctly :(

**Comment by renzhezizhi** [comment]
> @renzhezizhi can you comment on this PR about your finding -> #24155
> 
> Because I think that PR is trying to fix the issue in a similar way 🤗

Sure. Have done. XD

**Comment by omatrot** [comment]
Okay, I started from scratch with a new project. I still have the warnings but the app runs.

**Comment by mrTuomoK** [comment]
So was this fixed or not?

**Comment by kelset** [comment]
@mrTuomoK it looks like a PR with the fix landed (https://github.com/facebook/react-native/commit/ff66600224e78fec5d0e902f8a035b78ed31a961), I'll see if we can cherry-pick it for 0.59.3 but if it's not possible it will be out in 0.60

**Comment by githubdoramon** [comment]
@kelset are there any expectations on release dates (approximately) for 0.59.3 and 0.60?

**Comment by kelset** [comment]
@githubdoramon we just releases 0.59.3, for 0.60 it will be a few more weeks I feel, we are still working on one key feature we want to have in it (you can always keep an eye on -> https://github.com/react-native-community/react-native-releases)

Btw, pls LMK if now it's fixed in 0.59.3

**Comment by rjruizes** [comment]
Fixed in 0.59.3 for me (iOS 12.2, Xcode 10.2)

**Comment by githubdoramon** [comment]
@kelset Having this on 0.59.3 was more than enough. Thanks a lot... Also working here

**Comment by benevbright** [comment]
Anybody have not a problem with `NativeAppEventEmitter` in RN 0.53 after this patch? 
(I applied this patch on my old RN manually)

**Comment by benevbright** [comment]
Hm, actually `NativeModules.UserEventEmitter` becomes `undefined`.

**Comment by benevbright** [comment]
No, my bad. Sorry guys.
It's working fine.

**Comment by thepost** [comment]
I'm getting the same problem. I've narrowed it down to the point where you add an Obj-C implementation file wrapped in `RCT_EXTERN_MODULE`. This causes a crash. Adding the file itself does not cause a crash, it's not until you wrap the interface in `RCT_EXTERN_MODULE`.

This is consistent even when the class is empty.

**Comment by thepost** [comment]
> I've implemented a (temporary) workaround, but I'm not sure how to fix it properly in React.
> 
> In your (Swift based) bridge modules, use the marco `RCT_EXPORT_PRE_REGISTERED_MODULE()` instead of `RCT_EXPORT_MODULE`. This prevents the `+ (void)load` method to be implemented on it.
> 

Could you explain a little bit more about where you'd use `RCT_EXPORT_PRE_REGISTERED_MODULE` please? Using it in my .m file in the same way that you'd use `RCT_EXTERN_MODULE` has no affect. So do you use it in your Swift file, or a .m?

**Comment by RadoAz** [comment]
Hi, guys. I solved this problem adding this 12.2 folder ( https://github.com/iGhibli/iOS-DeviceSupport/blob/master/DeviceSupport/12.2%20(16E226).zip ) to Xcode(in applications)  -> Show Package Contents -> Contents -> Developer -> iPhoneIOs.platform -> Device Support (you will see list of folders: 8.0, 8.1, 8.2 etc...). After you will add folder named 12.2 here - everything will work fine!

**Comment by shreyasthiagaraj** [comment]
Still having this isssue, using react-native 0.59.8. Anyone else still experiencing this?
